profiles/api/oauth2.py

```diff
@@ -0,0 +1,139 @@
+from json import JSONDecodeError, dumps
+from urllib.parse import urlencode
+import requests
+from flask import Blueprint, current_app, make_response, redirect, request, jsonify, url_for, json
+from profiles.api.errors import InvalidClient, InvalidRequest, UnsupportedGrantType, InvalidGrant, \
+    ClientInvalidRequest, ClientUnsupportedResourceType, ClientInvalidScope
+from profiles.utilities import remove_none_values
+from werkzeug.exceptions import BadRequest
+from werkzeug.wrappers import Response
+
+OAUTH2_BP = Blueprint('oauth', __name__)
+
+
+def find_client_by_id(client_id: str) -> dict:
+    clients = current_app.config['config']['oauth2']['clients']
+
+    try:
+        (name, details) = next((k, v) for k, v in clients.items() if v['id'] == client_id)
+    except StopIteration:
+        raise KeyError('No client with the client_id')
+
+    details['name'] = name
+
+    return details
+
+
+@OAUTH2_BP.route('/authorize')
+def authorize() -> Response:
+    if 'client_id' not in request.args:
+        raise BadRequest('Invalid client_id')
+
+    try:
+        client = find_client_by_id(request.args.get('client_id'))
+    except KeyError:
+        raise BadRequest('Invalid client_id')
+
+    if request.args.get('redirect_uri', client['redirect_uri']) != client['redirect_uri']:
+        raise BadRequest('Invalid redirect_uri')
+
+    if not request.args.get('response_type'):
+        raise ClientInvalidRequest(client, 'Missing response_type')
+    elif request.args.get('response_type') != 'code':
+        raise ClientUnsupportedResourceType(client)
+
+    if request.args.get('scope'):
+        raise ClientInvalidScope(client)
+
+    state = remove_none_values({
+        'redirect_uri': client['redirect_uri'],
+        'client_id': client['id'],
+        'original': request.args.get('state')
+    })
```

Comment:
(Restoring still-valid comment.)

Rather than having to pass all this, might be better to store locally and just include the key. Also gives CSRF protection for free (if random).

Reply:
Maybe even store it in a session, rather than persisting it in a DB or key-value store?

Reply:
Adding to the options: what about encrypting or signing it if it solves CSRF protection, without the persistence?

Reply:
Could sign it, but still it's getting quite a bit of data exposed in the URI. Let's create a ticket for it and do it later.